src/jit/optimizer.cpp

```diff
@@ -1416,410 +1418,1018 @@ void Compiler::optCheckPreds()
 
 #endif // DEBUG
 
-/*****************************************************************************
- * Find the natural loops, using dominators. Note that the test for
- * a loop is slightly different from the standard one, because we have
- * not done a depth first reordering of the basic blocks.
- */
-
-void Compiler::optFindNaturalLoops()
+namespace
 {
-#ifdef DEBUG
-    if (verbose)
+//------------------------------------------------------------------------
+// LoopSearch: Class that handles scanning a range of blocks to detect a loop,
+//             moving blocks to make the loop body contiguous, and recording
+//             the loop.
+//
+// We will use the following terminology:
+//   HEAD    - the basic block that flows into the loop ENTRY block (Currently MUST be lexically before entry).
+//             Not part of the looping of the loop.
+//   FIRST   - the lexically first basic block (in bbNext order) within this loop.  (May be part of a nested loop,
+//             but not the outer loop. ???)
```

Comment:
Thanks for prompting me to take a closer look at this.  I had misread and thought the old code wouldn't identify loops that shared a `TOP` or `ENTRY` or `FIRST` -- on closer read, it allowed it as long as `entry`'s predecessor list had an inner-loop predecessor prior to having an outer-loop predecessor (oddly).  So I've fixed my code to recognize nested loops that share `TOP`/`FIRST`/`ENTRY` in https://github.com/dotnet/coreclr/pull/13314/commits/13bcb530f4f66ca9bdd006fab7a2d7c7f3040c97 (before refactoring) and https://github.com/dotnet/coreclr/pull/13314/commits/1091e8947ac147d2a900aa0055909402566c5704 + https://github.com/dotnet/coreclr/pull/13314/commits/a9c20d2ec763186650e1c0ff49c572760e5d7525 (after refactoring), without the bizarro restriction on predecessor list.